snuba/migrations/validator.py

```diff
@@ -0,0 +1,194 @@
+import re
+from typing import Sequence, Union
+
+from snuba.clickhouse.native import ClickhousePool
+from snuba.clusters.cluster import ClickhouseClientSettings, get_cluster
+from snuba.migrations.migration import ClickhouseNodeMigration
+from snuba.migrations.operations import AddColumn, CreateTable, DropColumn, SqlOperation
+
+ENGINE_REGEX = r"^Distributed(\(.+\))$"
+
+
+class InvalidMigrationOrderError(Exception):
+    """
+    Raised when a migration ops are not in the correct order.
+    """
+
+    pass
+
+
+class NoDistClusterNodes(Exception):
+    """
+    The migration has no distributed cluster nodes, so we can't validate the order.
+    """
+
+    pass
+
+
+class DistributedEngineParseError(Exception):
+    """
+    Cant parser the distributed engine string value
+    """
+
+    pass
+
+
+def validate_migration_order(migration: ClickhouseNodeMigration) -> None:
+    """
+    Validates that the migration order is correct. Chesk that the order of
+    AddColumn, CreateTable and DropColumn operations are correct with reagards to being
+    applied on local and distributed tables.
+    """
+
+    def validate_order(
+        local_ops: Sequence[SqlOperation],
+        dist_ops: Sequence[SqlOperation],
+        local_first: bool,
+    ) -> None:
+        if local_first:
+            for dist_op in dist_ops:
+                if isinstance(dist_op, DropColumn):
```

Comment:
I think we can make some of this logic a bit more DRY by pulling out the logic for validating local and dist ops

```python
def validate_dist_op(dist_op, local_ops, operation) -> None:
    if not isinstance(dist_op, operation):
        return
    for local_op in local_ops:
        if not isinstance(local_op, operation):
            return
        if conflicts_ops(operation, local_op, dist_op):
            raise InvalidMigrationOrderError(
                f"{operation} {dist_op} \
                operation must be applied on dist table before local"
            )


def validate_local_op(local_op, dist_ops, operation) -> None:
    if not isinstance(local_op, operation):
        return
    for dist_op in dist_ops:
        if not isinstance(dist_op, operation):
            return
        if conflicts_ops(operation, local_op, dist_op):
            raise InvalidMigrationOrderError(
                f"{operation} {dist_op} \
                operation must be applied on local table before dist"
            )
```

You'd have to have some way to repr the operation as a string depending on the op but
this would all you to write something like 

```python
def validate_order(
        local_ops: Sequence[SqlOperation],
        dist_ops: Sequence[SqlOperation],
        local_first: bool,
    ) -> None:
        if local_first:
            for dist_op in dist_ops:
                validate_dist_op(dist_op, local_ops, DropColumn)

        else:
            for local_op in local_ops:
                validate_local_op(local_op, dist_ops, AddColumn)
                validate_local_op(local_op, dist_ops, CreateTable)
```

You'd need to unify the logic for the conflicts function, but you could have mapping from operation to callable or possibly there is a way to make the conflicts_ops encapsulate the logic you have for all three methods right now. They look pretty similar except for create table one

Reply:
Agreed. I have refactored it to be more DRY now. The only issue is the type checker complains so I had to have a few more conditionals to check op type when inside the same function.